lib/google/cloud/gemserver/authentication.rb

```diff
@@ -52,12 +56,126 @@ def can_modify?
           editors.each do |editor|
             return true if extract_account(editor) == user
           end
+          puts "You are either not authenticated with gcloud or lack access" \
+            " to the gemserver."
           false
         end
 
+        ##
+        # Generates an access token from a user authenticated by gcloud.
+        #
+        # @return [String]
+        def access_token
+          return unless can_modify?
+          scope = ["https://www.googleapis.com/auth/cloud-platform"]
+          auth = Google::Auth.get_application_default scope
+          auth.fetch_access_token!
+        end
+
+        ##
+        # Uses the tokeninfo API to validate the given access token.
+        #
+        # @param [String] auth_header The authorization header containing the
+        # token, e.g. "Bearer [TOKEN]".
+        #
+        # @return [Boolean]
+        def validate_token auth_header
+          token = auth_header.split.drop(1)[0]
+          apis_url = "https://www.googleapis.com"
+          tokeninfo_endpoint = "/oauth2/v1/tokeninfo?access_token=#{token}"
+          res = send_req apis_url, tokeninfo_endpoint, Net::HTTP::Post, token
+
+          return false unless check_status(res)
+
+          token_can_edit? token
+        end
+
         private
 
         ##
+        # @private Implicitly checks if the account that generated the token
+        # has edit permissions on the Google Cloud Platform project by issuing
+        # a redundant update to the project (update to original settings).
+        #
+        # @param [String] The authentication token generated from gcloud.
+        #
+        # @return [Boolean]
+        def token_can_edit? token
+          appengine_url = "https://appengine.googleapis.com"
+          endpoint = "/v1/apps/#{@proj}/services/default?updateMask=split"
+          version = appengine_version token
+          split = {
+            "split" => {
+              "allocations" => {
+                version.to_s => 1
+              }
+            }
+          }
+          res = send_req appengine_url, endpoint, Net::HTTP::Patch, token, split
+          check_status res
+        end
+
+        ##
+        # @private Fetches the latest version of the deployed Google App Engine
+        # instance running the gemserver (default service only).
+        #
+        # @param [String] The authentication token generated from gcloud.
+        #
+        # @return [String]
+        def appengine_version token
+          appengine_url = "https://appengine.googleapis.com"
+          path = "/v1/apps/#{@proj}/services/default"
+          res = send_req appengine_url, path, Net::HTTP::Get, token
+
+          fail "Unauthorized" unless check_status(res)
+
+          eval(res.body)[:split][:allocations].first[0]
```

Comment:
Is it JSON? (I assume it's not actually a Ruby hash.) Can we just use a JSON parser? Eval is dangerous.